### PR TITLE
set default encoding of written files to UTF-8, added ability to pass custom file encoding

### LINKED ISSUE #4302: On windows user can't write unicode symbols to file
SCons: v4.4.0.fc8d0ec215ee6cba8bc158ad40c099be0b598297 (installed via scoop)
Python 3.11.0 (default distribution)
Windows 11

Bug description: On windows user can't write unicode symbols to file, because default encoding is cp1251

This can be easily fixed by adding utf-8 encoding here
https://github.com/SCons/scons/blob/440728dd1d9fee6a8e010b4d9871737686cb3afb/SCons/Tool/textfile.py#L122

```
Traceback (most recent call last):
  File "C:\Users\Redwan\scoop\apps\scons\4.4.0\scons-local-4.4.0\SCons\Action.py", line 1318, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Redwan\scoop\apps\scons\4.4.0\scons-local-4.4.0\SCons\Tool\textfile.py", line 132, in _action
    target_file.write(_do_subst(line, subs))
  File "C:\Program Files\Python311\Lib\encodings\cp1251.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeEncodeError: 'charmap' codec can't encode character '\xd7' in position 8177: character maps to <undefined>
scons: building terminated because of errors.
[Time elapsed: 00:00:02.645]
```

Way to reproduce:
Try to write unicode symboles using `env.Textfile` on windows .

### PR DESCRIPTION
**Pull request by RedwanFox**
Fixes https://github.com/SCons/scons/issues/4302

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: SCons/Tool/textfile.py [hunk 1]
```diff
@@ -117,9 +117,14 @@ def _action(target, source, env):
                 value = str(value)
             subs.append((k, value))
 
+    if 'FILE_ENCODING' not in env:
+        file_encoding = 'utf-8'
+    else:
+        file_encoding = env['FILE_ENCODING']
```
Comment (bdbaddog) [inline SCons/Tool/textfile.py]: Please change the variable name to `TEXTFILE_FILE_ENCODING`

File: CHANGES.txt [hunk 2]
```diff
@@ -134,6 +134,11 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
       or not if it is already in the $LIBS construction var in the
       configure context. (issue #2768).
 
+  From Nickolai Korshunov
```
Comment (bdbaddog) [inline CHANGES.txt]: Please add your blurb to this section alphabetically ordered by your last name. (That's our practice here).

**Comment by bdbaddog** [comment]
Good work!

Please add a blurb to RELEASE.txt as well and address the couple comments and we should able to merge this quickly. (assuming no other regressions fail).

**Comment by mwichmann** [comment]
No interest in holding this up, but there are other places in the code where we also see this wanrning: 
```
SCons/Tool/textfile.py:122:22: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
```

So, just wanted to point out that limiting the construction var to `textfile` might or might not end up being the preferred approach long-term.

**Comment by bdbaddog** [comment]
> No interest in holding this up, but there are other places in the code where we also see this wanrning:
> 
> ```
> SCons/Tool/textfile.py:122:22: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
> ```
> 
> So, just wanted to point out that limiting the construction var to `textfile` might or might not end up being the preferred approach long-term.

Sure.
But since this change only affects textfile, the prefix should be ok.
Please if we add a global default (which affects all builders in an environment), then textfile could default to that if TEXTFILE_FILE_ENCODING isn't set, and additionally it may make sense for some to set such explicitly per text file. (Multi-language text distributions?)

**Comment by bdbaddog** [comment]
The changed test fails on linux and macos..

**Comment by bdbaddog** [comment]
Never mind on changing variable name. I have local updates to this PR I'll push in a few.

**Comment by RedwanFox** [comment]
>The changed test fails on linux and macos..

Macos fails on tests independent to this change as far as I can see (test/ninja/ninja_handle_control_c_rebuild.py, test/ninja/shutdown_scons_daemon.py).
I can't reproduce failing of linux tests locally because all tests explode under WSL for me. 

@bdbaddog  do I need to push anything after your patch?

**Comment by bdbaddog** [comment]
> > The changed test fails on linux and macos..
> 
> Macos fails on tests independent to this change as far as I can see (test/ninja/ninja_handle_control_c_rebuild.py, test/ninja/shutdown_scons_daemon.py). I can't reproduce failing of linux tests locally because all tests explode under WSL for me.
> 
> @bdbaddog do I need to push anything after your patch?

Your test failed on linux and macos. (Not just that the test suite had any failures).
No need to change after my patch.
It looks like you forgot to commit/change the fixture's SConstruct to include the utf-8 only legal character.
After adding that the test now passes on all tested platforms.

**Comment by RedwanFox** [comment]
Great, thanks!

**Comment by bdbaddog** [comment]
Ideally we'd add a test for a non-default (not utf-8) target file as well. To ensure that setting FILE_ENCODING propagates and affects the written target file.
I'm not sure how best to do that though.
Are you familiar with any non-utf-8 encodings which would be reasonable to test?

**Comment by RedwanFox** [comment]
Cyrillic symbols cp1251 encoding. Or japanese encodings though for them I can't give you names or examples.

**Comment by bdbaddog** [comment]
Hmmm. could set it to ascii and then try to write the same data and expect a failure/exception..

**Pull request merged**